### DISCUSSION TITLE
[suggestion] Always-in-sync fork repos for maintaining external PRs

### DISCUSSION THREAD

**Discussion by tsutsu** [community/community#19566] [Other Feature Feedback, Questions, & Ideas]
If I don't have commit rights to an upstream repo, and I want to submit a PR to said repo, then I have to fork that repo.

And if my PR lives a long time, I might want to rebase my fork against the upstream to ensure it stays mergeable.

In fact, I want all changes to the upstream to be "inherited" by my fork. Because I don't really want to have a _fork_; I just want to have a feature-branch that lives in the upstream repo, if they'd let me! I want a workflow that _simulates_, as closely as possible, the semantics of having my PR be based on a feature branch that lives in the upstream repo.

Despite this being a major use-case of repo forking on Github, forks don't support this use-case very well. The design for forks seems to assume that you want to maintain your fork entirely separately from the upstream it was forked from.

It would be much nicer if repos forked "for PR submission purposes", a.k.a. "external PR branches", could be their own first-class Github model-objects, with their own UI and semantics tuned for this common use-case.

These forks should:
- be hidden from view under the user's regular Repositories view, with just the branch or PR itself surfaced instead in a per-user/per-org "External PRs" view (where you can then navigate to the implicit fork-repo from the PR)
- be named with a generated name so as to never collide with any repo the user creates explicitly
- be set up so that all existing branches and tags of the parent repo are _mirrored_ into the fork, always-in-sync with the parent repo and read-only, with the fork owner only able to create new branches
- restrict (or auto-decorate) the names of branches created in them to a certain naming scheme, that effectively namespaces them to the user or org that owns them, so that mirrored parent branches will never collide with them; and, likewise, restrict github-hosted repos generally from having branches of this naming pattern (so that parents can never intentionally create a branch that collides with a later fork branch)
- "pass through" all Github model features (e.g. Issues, PRs, etc.) on the fork repo to the parent repo — making the fork repo act as an "overlay" or "proxy" for the parent repo
- in the web UI, surface the fork-repo's new branches as "virtual" branches _of_ the parent repo

Obviously, the Github "Edit" button would then create these "external PR branches", rather than creating named non-mirroring fork repos as it does now.

**Comment by JohnVeness** [comment]
The [Pull app](https://wei.github.io/pull/) will do this, although I agree it would be good if this was a built-in GH feature.
